raiden/transfer/mediated_transfer/mediator.py

```diff
@@ -224,60 +226,70 @@ def get_pending_transfer_pairs(
-    channel: NettingChannelState, amount: PaymentWithFeeAmount
-) -> Optional[FeeAmount]:
-    """ Fee deducted by the mediator for an outgoing channel.
-
-    The `amount` is the incoming amount where the incoming fee is already
-    deducted, but the outgoing fee isn't (that's what this function does,
-    after all).
-    """
-    balance = get_balance(channel.our_state, channel.partner_state)
+    i = 0
+    y = fee_func.y_list[i]
+    comp = operator.lt if y < line(i) else operator.gt
+    while comp(y, line(i)):
+        i += 1
+        if i == len(fee_func.x_list):
+            # Not enough capacity to send
+            return None
+        y = fee_func.y_list[i]
+
+    # We found the linear section where the solution is. Now interpolate!
+    x1 = fee_func.x_list[i - 1]
+    x2 = fee_func.x_list[i]
+    yf1 = fee_func.y_list[i - 1]
+    yf2 = fee_func.y_list[i]
+    yl1 = line(i - 1)
+    yl2 = line(i)
+    return (yl1 - yf1) * (x2 - x1) / ((yf2 - yf1) - (yl2 - yl1)) + x1
+
+
+def get_amount_without_fees(
+    amount_with_fees: PaymentWithFeeAmount,
+    channel_in: NettingChannelState,
+    channel_out: NettingChannelState,
+) -> Optional[PaymentWithFeeAmount]:
+    """ Return the amount after fees are taken. """
 
+    balance_in = get_balance(channel_in.our_state, channel_in.partner_state)
+    balance_out = get_balance(channel_out.our_state, channel_out.partner_state)
+    receivable = TokenAmount(
+        channel_in.our_total_deposit + channel_in.partner_total_deposit - balance_in
+    )
+    assert (
+        channel_in.fee_schedule.cap_fees == channel_out.fee_schedule.cap_fees
+    ), "Both channels must have the same cap_fees setting for the same mediator."
     try:
-        return channel.fee_schedule.fee_payee(amount=amount, balance=balance)
+        fee_func = FeeScheduleState.mediation_fee_func(
+            schedule_in=channel_in.fee_schedule,
+            schedule_out=channel_out.fee_schedule,
+            balance_in=balance_in,
+            balance_out=balance_out,
+            receivable=receivable,
+            amount_with_fees=amount_with_fees,
+            cap_fees=channel_in.fee_schedule.cap_fees,
+        )
+        amount_without_fees = find_intersection(
+            fee_func, lambda i: amount_with_fees - fee_func.x_list[i]
+        )
     except UndefinedMediationFee:
         return None
 
-
-def get_lock_amount_after_fees(
-    lock: HashTimeLockState, payer_channel: NettingChannelState, payee_channel: NettingChannelState
-) -> Optional[PaymentWithFeeAmount]:
-    """
-    Return the lock.amount after fees are taken.
-
-    Fees are taken only for the outgoing channel, which is the one with
-    collateral locked from this node.
-    """
-    fee_in = _fee_for_payer_channel(payer_channel, lock.amount)
-    if fee_in is None:
-        return None
-    fee_out = _fee_for_payee_channel(payee_channel, PaymentWithFeeAmount(lock.amount - fee_in))
-    if fee_out is None:
+    if amount_without_fees is None:
```

Comment:
Can you add a comment here that this means that we don't have enough capacity to send? This would make it easier for someone to understand the logic without jumping around too much.